core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/handler/BaseClientHandler.java

```diff
@@ -182,7 +197,67 @@ protected boolean isCalculateCrc32FromCompressedData() {
+    }
+
+    <OutputT extends SdkResponse> HttpResponseHandler<Response<OutputT>> decorateSuccessResponseHandlers(
+        HttpResponseHandler<Response<OutputT>> delegate, ExecutionContext executionContext) {
+
+        return successTransformationResponseHandler(delegate, responseTransformations(executionContext));
+    }
+
+    <OutputT extends SdkResponse> HttpResponseHandler<Response<OutputT>> successTransformationResponseHandler(
+        HttpResponseHandler<Response<OutputT>> responseHandler,
+        BiFunction<OutputT, SdkHttpFullResponse, OutputT> successTransformer) {
+
+        return (response, executionAttributes) -> {
+            Response<OutputT> delegateResponse = responseHandler.handle(response, executionAttributes);
+
+            if (delegateResponse.isSuccess()) {
+                return delegateResponse.toBuilder()
+                                       .response(successTransformer.apply(delegateResponse.response(), response))
+                                       .build();
+            } else {
+                return delegateResponse;
+            }
+        };
+    }
+
+    <OutputT extends SdkResponse> HttpResponseHandler<OutputT> resultTransformationResponseHandler(
+        HttpResponseHandler<OutputT> responseHandler,
+        BiFunction<OutputT, SdkHttpFullResponse, OutputT> successTransformer) {
+
+        return (response, executionAttributes) -> {
+            OutputT delegateResponse = responseHandler.handle(response, executionAttributes);
+            return successTransformer.apply(delegateResponse, response);
+        };
+    }
+
+
+    static void validateExecutionParams(ClientExecutionParams<?, ?> executionParams) {
+        if (executionParams.getCombinedResponseHandler() != null) {
+            if (executionParams.getResponseHandler() != null) {
+                throw new IllegalArgumentException("Only one of 'combinedResponseHandler' and 'responseHandler' may "
+                                                   + "be specified in a ClientExecutionParams object");
+            }
+
+            if (executionParams.getErrorResponseHandler() != null) {
+                throw new IllegalArgumentException("Only one of 'combinedResponseHandler' and 'errorResponseHandler' "
+                                                   + "may be specified in a ClientExecutionParams object");
+            }
+        }
+    }
+
+    private static <T extends SdkResponse> BiFunction<T, SdkHttpFullResponse, T>
+        responseTransformations(ExecutionContext executionContext) {
+
+        return composeResponseFunctions(runAfterUnmarshallingInterceptors(executionContext),
+                                        attachHttpResponseToResult());
+    }
+
+    private static <T, R> BiFunction<T, R, T> composeResponseFunctions(BiFunction<T, R, T> function1,
+                                                               BiFunction<T, R, T> function2) {
```

Comment:
Wouldn't think this passes checkstyle either...

Reply:
Why not?

Reply:
The second BiFunction should line up with the first.

Reply:
Will fix.